File: packages/js/core/src/memory.js
export class MemoryEntry {
  constructor({ role, content, metadata = {}, createdAt = null } = {}) {
    this.role = role;
    this.content = content;
    this.metadata = metadata;
    this.createdAt = createdAt || new Date().toISOString();
  }

  toDict() {
    return {
      role: this.role,
      content: this.content,
      metadata: this.metadata,
      created_at: this.createdAt,
    };
  }

  static fromDict(data) {
    return new MemoryEntry({
      role: data.role || "",
      content: data.content || "",
      metadata: data.metadata || {},
      createdAt: data.created_at || data.createdAt,
    });
  }
}

export class AgentMemory {
  constructor({ entries = [] } = {}) {
    this.entries = entries.map(item => item instanceof MemoryEntry ? item : MemoryEntry.fromDict(item));
  }

  add(role, content, metadata = {}) {
    const entry = new MemoryEntry({ role, content, metadata });
    this.entries.push(entry);
    return entry;
  }

  last(count = 5) {
    if (count <= 0) return [];
    return this.entries.slice(-count);
  }

  toText(count = 5) {
    return this.last(count).map(entry => `${entry.role}: ${entry.content}`).join("\n");
  }

  clear() {
    this.entries = [];
  }
}

export class MemoryItem {
  constructor({ id = null, content, kind = "note", tags = [], metadata = {}, createdAt = null, updatedAt = null } = {}) {
    this.id = id || Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
    this.content = content;
    this.kind = kind;
    this.tags = Array.isArray(tags) ? tags : [];
    this.metadata = metadata;
    this.createdAt = createdAt || new Date().toISOString();
    this.updatedAt = updatedAt;
  }

  toDict() {
    return {
      id: this.id,
      content: this.content,
      kind: this.kind,
      tags: this.tags,
      metadata: this.metadata,
      created_at: this.createdAt,
      updated_at: this.updatedAt,
    };
  }

  toJS() {
    return this.toDict();
  }

  static fromDict(data) {
    return new MemoryItem({
      id: data.id,
      content: data.content || "",
      kind: data.kind || "note",
      tags: data.tags || [],
      metadata: data.metadata || {},
      createdAt: data.created_at || data.createdAt,
      updatedAt: data.updated_at || data.updatedAt,
    });
  }
}

export class MemoryStore {
  constructor({ items = [] } = {}) {
    this._items = new Map();
    for (const item of items) {
      const inst = item instanceof MemoryItem ? item : MemoryItem.fromDict(item);
      this._items.set(inst.id, inst);
    }
  }

  add(content, { kind = "note", tags = [], metadata = {} } = {}) {
    const item = new MemoryItem({ content, kind, tags, metadata });
    this._items.set(item.id, item);
    this._save();
    return item;
  }

  list() {
    return Array.from(this._items.values());
  }

  get(memoryId) {
    return this._items.get(memoryId) || null;
  }

  search(query, { limit = null } = {}) {
    if (!query) return [];
    const terms = query.toLowerCase().split(/\s+/).filter(Boolean);
    const scored = [];
    for (const item of this._items.values()) {
      const haystack = [
        item.content,
        item.kind,
        item.tags.join(" "),
        JSON.stringify(item.metadata),
      ].join(" ").toLowerCase();
      
      let score = 0;
      for (const term of terms) {
        let idx = -1;
        while ((idx = haystack.indexOf(term, idx + 1)) !== -1) {
          score++;
        }
      }
      
      if (score > 0) {
        scored.push({ score, item });
      }
    }
    
    // Sort: score descending, then createdAt ascending, then id ascending
    scored.sort((a, b) => {
      if (b.score !== a.score) return b.score - a.score;
      if (a.item.createdAt !== b.item.createdAt) return a.item.createdAt.localeCompare(b.item.createdAt);
      return a.item.id.localeCompare(b.item.id);
    });
    
    const results = scored.map(entry => entry.item);
    return limit !== null ? results.slice(0, limit) : results;
  }

  delete(memoryId) {
    const existed = this._items.delete(memoryId);
    if (existed) {
      this._save();
    }
    return existed;
  }

  clear() {
    this._items.clear();
    this._save();
  }

  _save() {
    // In-memory no-op
  }
}

export class MemoryReport {
  constructor({ memoriesStored = [], searchesExecuted = [], contextDocumentsUsed = [], finalResult = "" } = {}) {
    this.memoriesStored = memoriesStored.map(item => item instanceof MemoryItem ? item : MemoryItem.fromDict(item));
    this.searchesExecuted = searchesExecuted;
    this.contextDocumentsUsed = contextDocumentsUsed;
    this.finalResult = finalResult;
  }

  toDict() {
    return {
      memories_stored: this.memoriesStored.map(item => item.toDict()),
      searches_executed: this.searchesExecuted,
      context_documents_used: this.contextDocumentsUsed,
      final_result: this.finalResult,
    };
  }

  toJS() {
    return this.toDict();
  }

  toMarkdown() {
    const memories = this.memoriesStored.map(item => `- ${item.kind}: ${item.content}`).join("\n");
    const searches = this.searchesExecuted.map(item => `- ${item}`).join("\n");
    const docs = this.contextDocumentsUsed.map(item => `- ${item}`).join("\n");
    return [
      "# Memory Report",
      "",
      "## Memories Stored",
      "",
      memories || "- none",
      "",
      "## Searches Executed",
      "",
      searches || "- none",
      "",
      "## Context Documents Used",
      "",
      docs || "- none",
      "",
      "## Final Result",
      "",
      this.finalResult,
    ].join("\n") + "\n";
  }
}

// ==========================================
// Extensions Registry
// ==========================================

File: python/packages/handoffkit/handoffkit/memory.py
"""Agent memory plus structured memory stores."""

from __future__ import annotations

import json
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any


def _now() -> str:
    """Return current UTC timestamp."""
    return datetime.now(timezone.utc).isoformat()


@dataclass
class MemoryEntry:
    """One memory entry created by an agent run."""

    role: str
    content: str
    metadata: dict[str, Any] = field(default_factory=dict)
    created_at: str = field(default_factory=_now)


@dataclass
class AgentMemory:
    """Small append-only memory used by agents."""

    entries: list[MemoryEntry] = field(default_factory=list)

    def add(self, role: str, content: str, **metadata: Any) -> MemoryEntry:
        """Add an entry and return it."""
        entry = MemoryEntry(role=role, content=content, metadata=metadata)
        self.entries.append(entry)
        return entry

    def last(self, count: int = 5) -> list[MemoryEntry]:
        """Return the most recent entries."""
        if count <= 0:
            return []
        return self.entries[-count:]

    def to_text(self, count: int = 5) -> str:
        """Render recent memory entries as compact text."""
        return "\n".join(f"{entry.role}: {entry.content}" for entry in self.last(count))

    def clear(self) -> None:
        """Remove all entries."""
        self.entries.clear()


@dataclass
class MemoryItem:
    """Structured memory item stored across agent runs."""

    content: str
    kind: str = "note"
    tags: list[str] = field(default_factory=list)
    metadata: dict[str, Any] = field(default_factory=dict)
    id: str = field(default_factory=lambda: uuid.uuid4().hex)
    created_at: str = field(default_factory=_now)
    updated_at: str | None = None

    def to_dict(self) -> dict[str, Any]:
        """Serialize this memory item."""
        return {
            "id": self.id,
            "content": self.content,
            "kind": self.kind,
            "tags": self.tags,
            "metadata": self.metadata,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
        }

    def to_json(self, *, indent: int | None = 2) -> str:
        """Serialize this memory item as JSON."""
        return json.dumps(self.to_dict(), ensure_ascii=False, indent=indent)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> MemoryItem:
        """Create a memory item from a dictionary."""
        return cls(
            id=str(data.get("id") or uuid.uuid4().hex),
            content=str(data.get("content", "")),
            kind=str(data.get("kind", "note")),
            tags=list(data.get("tags") or []),
            metadata=dict(data.get("metadata") or {}),
            created_at=str(data.get("created_at") or _now()),
            updated_at=data.get("updated_at"),
        )

    @classmethod
    def from_json(cls, value: str) -> MemoryItem:
        """Create a memory item from JSON."""
        data = json.loads(value)
        if not isinstance(data, dict):
            raise ValueError("MemoryItem JSON must decode to an object.")
        return cls.from_dict(data)


class MemoryStore:
    """Deterministic in-memory store for structured memories."""

    def __init__(self, items: list[MemoryItem] | None = None) -> None:
        self._items: dict[str, MemoryItem] = {}
        for item in items or []:
            self._items[item.id] = item

    def add(
        self,
        content: str,
        *,
        kind: str = "note",
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> MemoryItem:
        """Add one memory item."""
        item = MemoryItem(
            content=content,
            kind=kind,
            tags=tags or [],
            metadata=metadata or {},
        )
        self._items[item.id] = item
        self._save()
        return item

    def list(self) -> list[MemoryItem]:
        """Return all memory items in insertion order."""
        return list(self._items.values())

    def get(self, memory_id: str) -> MemoryItem | None:
        """Return one memory item by id."""
        return self._items.get(memory_id)

    def search(self, query: str, *, limit: int | None = None) -> list[MemoryItem]:
        """Search memories with deterministic keyword scoring."""
        terms = [term.lower() for term in query.split() if term.strip()]
        scored: list[tuple[int, MemoryItem]] = []
        for item in self._items.values():
            haystack = " ".join(
                [
                    item.content,
                    item.kind,
                    " ".join(item.tags),
                    json.dumps(item.metadata, default=str),
                ]
            ).lower()
            score = sum(haystack.count(term) for term in terms)
            if score > 0:
                scored.append((score, item))
        scored.sort(key=lambda entry: (-entry[0], entry[1].created_at, entry[1].id))
        results = [item for _, item in scored]
        return results[:limit] if limit is not None else results

    def delete(self, memory_id: str) -> bool:
        """Delete one memory item."""
        existed = self._items.pop(memory_id, None) is not None
        if existed:
            self._save()
        return existed

    def clear(self) -> None:
        """Delete all memory items."""
        self._items.clear()
        self._save()

    def _save(self) -> None:
        """Persist memory items. In-memory store is a no-op."""


class JsonMemoryStore(MemoryStore):
    """JSON-backed memory store."""

    def __init__(self, path: str) -> None:
        self.path = Path(path)
        self.path.parent.mkdir(parents=True, exist_ok=True)
        if not self.path.exists():
            self.path.write_text("[]", encoding="utf-8")
        items = self._load_items()
        super().__init__(items)

    def _load_items(self) -> list[MemoryItem]:
        raw = self.path.read_text(encoding="utf-8").strip()
        if not raw:
            return []
        try:
            data = json.loads(raw)
        except json.JSONDecodeError as exc:
            raise ValueError(f"Invalid JSON memory store: {self.path}") from exc
        if not isinstance(data, list):
            raise ValueError(f"JSON memory store must contain a list: {self.path}")
        return [MemoryItem.from_dict(item) for item in data if isinstance(item, dict)]

    def _save(self) -> None:
        self.path.write_text(
            json.dumps([item.to_dict() for item in self.list()], ensure_ascii=False, indent=2),
            encoding="utf-8",
        )


@dataclass
class MemoryReport:
    """Report describing memory/context usage."""

    memories_stored: list[MemoryItem] = field(default_factory=list)
    searches_executed: list[str] = field(default_factory=list)
    context_documents_used: list[str] = field(default_factory=list)
    final_result: str = ""

    def to_dict(self) -> dict[str, Any]:
        """Serialize report."""
        return {
            "memories_stored": [item.to_dict() for item in self.memories_stored],
            "searches_executed": self.searches_executed,
            "context_documents_used": self.context_documents_used,
            "final_result": self.final_result,
        }

    def to_json(self, *, indent: int | None = 2) -> str:
        """Serialize report as JSON."""
        return json.dumps(self.to_dict(), ensure_ascii=False, indent=indent)

    def to_markdown(self) -> str:
        """Serialize report as Markdown."""
        memories = "\n".join(f"- {item.kind}: {item.content}" for item in self.memories_stored)
        searches = "\n".join(f"- {item}" for item in self.searches_executed)
        docs = "\n".join(f"- {item}" for item in self.context_documents_used)
        return (
            "# Memory Report\n\n"
            f"## Memories Stored\n\n{memories or '- none'}\n\n"
            f"## Searches Executed\n\n{searches or '- none'}\n\n"
            f"## Context Documents Used\n\n{docs or '- none'}\n\n"
            f"## Final Result\n\n{self.final_result}\n"
        )
